### PR TITLE
Nav description v2

### PR DESCRIPTION
**Pull request by anoblet**

### CODE REVIEW

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 1]
```diff
@@ -273,7 +280,27 @@ export class CXLMarketingNavElement extends LitElement {
         );
 
         // Populate.
-        contextMenu.items = this._createContextMenuItems(menuItem.children);
+        const contextMenuItems = this._createContextMenuItems(menuItem.children);
+        contextMenu.addEventListener('opened-changed', () => {
+          contextMenu.items = contextMenuItems;
+          const listBox = document.querySelector('vaadin-context-menu-list-box');
+          setTimeout(() => {
+            const { width } = getComputedStyle(listBox);
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:288]: This is clever :+1: 

Curious, why do we have to be wrapped in a `setTimeout()`?

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 2]
```diff
@@ -273,7 +280,27 @@ export class CXLMarketingNavElement extends LitElement {
         );
 
         // Populate.
-        contextMenu.items = this._createContextMenuItems(menuItem.children);
+        const contextMenuItems = this._createContextMenuItems(menuItem.children);
+        contextMenu.addEventListener('opened-changed', () => {
+          contextMenu.items = contextMenuItems;
+          const listBox = document.querySelector('vaadin-context-menu-list-box');
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:286]: This is guaranteed to be unique? Any available opportunities to become more specific with the selector?

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 3]
```diff
@@ -273,7 +280,27 @@ export class CXLMarketingNavElement extends LitElement {
...
+          setTimeout(() => {
+            const { width } = getComputedStyle(listBox);
+            if (menuItem.description) {
+              const el = document.createElement('vaadin-context-menu-item');
+              el.style.minWidth = '256px';
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:291]: For text stuff, it usually makes sense to use `ch` units.

But I'm not sure min-width is a good idea wrt to x-device compatibility. Let's deal with an iteration when real content setup demands it at one point.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 4]
```diff
@@ -273,7 +280,27 @@ export class CXLMarketingNavElement extends LitElement {
...
+              el.style.maxWidth = width;
+              el.classList.add('description');
+              render(html` ${unsafeHTML(menuItem.description)} `, el);
+              contextMenu.items = [
+                ...contextMenuItems,
+                { component: document.createElement('hr') },
+                { component: el },
+              ];
+              contextMenu.render();
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:300]: Hm, we manually render now?

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 2 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:286]: The top level menu will always be first in the document, I do not know a way to add any specificity such as id or class to `vaadin-context-menu-list-box`.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 1 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:288]: From what I observed, without `setTimeout` the menu wasn't finished rendering and the width calculation was incorrect.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 3 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:291]: I agree with you on using `ch` units. The code needs to be refactored to account for a switch in device orientation. I can work on that.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 4 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:300]: Without `contextMenu.render()` the description isn't added. I believe context menu does it's calculations before the `opened-changed` event.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 1 above)
Comment (lkraav) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:288]: So the primary complication is that we have to always add the "context content" menu item after initial menu open - just so we can get the initial width?

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:288]: Correct

**Comment by anoblet** [comment]
* Refactored for cleanup
* Accommodate `vaadin-context-menu` min-width/min-height restrictions

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 5]
```diff
@@ -273,7 +280,48 @@ export class CXLMarketingNavElement extends LitElement {
         );
 
         // Populate.
-        contextMenu.items = this._createContextMenuItems(menuItem.children);
+        const contextMenuItems = this._createContextMenuItems(menuItem.children);
+        contextMenu.items = contextMenuItems;
+
+        const setDescription = async () => {
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:286]: Curious, would refactoring this -> "private" `_setContextMenuItemDescription()` class method make sense / is technically it feasible?

It's quite a large body in the middle of already a large method.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 6]
```diff
@@ -273,7 +280,48 @@ export class CXLMarketingNavElement extends LitElement {
...
+          });
+
+          const observe = (mediaQueryString, callback) => {
+            const observer = window.matchMedia(mediaQueryString);
+            const matches = (mediaQueryList) => callback(mediaQueryList.matches);
+            matches(observer);
+            observer.addEventListener('change', matches);
+          };
+
+          observe('(min-width: 420px) and (min-height: 420px)', () => {
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:320]: This magic number is from upstream vaadin-context-menu internals? We can't refer to it via some symbol?

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 7]
```diff
@@ -273,7 +280,48 @@ export class CXLMarketingNavElement extends LitElement {
...
+              { component: document.createElement('hr') },
+              { component: description },
+            ];
+            contextMenu.render();
+          }, 0);
+        };
+
+        if (menuItem.description) {
+          contextMenu.addEventListener('opened-changed', () => {
+            if (!document.querySelector('vaadin-context-menu-list-box')) return;
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:309]: Wouldn't regular positive check be more sane here?

```js
             if (document.querySelector('vaadin-context-menu-list-box')) {
               setDescription();
             }
```

We're looking to get away from multiple statements on single line, maybe the ESLint standard hasn't been included.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 8]
```diff
@@ -273,7 +280,48 @@ export class CXLMarketingNavElement extends LitElement {
...
+
+          const observe = (mediaQueryString, callback) => {
+            const observer = window.matchMedia(mediaQueryString);
+            const matches = (mediaQueryList) => callback(mediaQueryList.matches);
+            matches(observer);
+            observer.addEventListener('change', matches);
+          };
+
+          observe('(min-width: 420px) and (min-height: 420px)', () => {
+            if (!document.querySelector('vaadin-context-menu-list-box')) return;
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:321]: Same as above.

**Review by lkraav — CHANGES_REQUESTED** [review]
This looks promising :+1:

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 5 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:286]: This makes sense.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 6 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:320]: I'm not able to find any symbols.

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 7 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:309]: Agree

File: packages/cxl-ui/src/components/cxl-marketing-nav.js (hunk 8 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:321]: Agree

**Comment by anoblet** [comment]
Updated to accommodate review suggestions

File: packages/cxl-ui/src/components/cxl-marketing-nav.js [hunk 9]
```diff
@@ -273,7 +275,29 @@ export class CXLMarketingNavElement extends LitElement {
...
+          });
+
+          const observe = (mediaQueryString, callback) => {
+            const observer = window.matchMedia(mediaQueryString);
+            const matches = (mediaQueryList) => callback(mediaQueryList.matches);
+            matches(observer);
+            observer.addEventListener('change', matches);
+          };
+
+          observe('(min-width: 420px) and (min-height: 420px)', () => {
```
Comment (lkraav) [inline packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: Curious, why do we have to have a seperate description operation for mobile at all - does the context menu get rebuilt? I think I've never had to react to anything on media change for this component, so not sure why it's useful.

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: `vaadin-context-menu` relies on a media change of `420px`

Comment (lkraav) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: Yes, but why wouldn't just populating description once on `opened-changed` event work universally?

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: User on mobile vertical decides to switch to horizontal. Continuity of product.

Comment (lkraav) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: Ahhhh, right, it's about the width update, not content - brain must've been fried, not sure how I missed that.

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-marketing-nav.js:295]: My brain was a little fried too, otherwise I would have been clearer in my explanation :)

**Comment by kylebrodeur** [comment]
Task linked: [CU-2x7c7dv Aybolit - Nav Menu items `description` feature.](https://app.clickup.com/t/2x7c7dv)

**Comment by saas786** [comment]
closed in favour of https://github.com/conversionxl/aybolit/pull/220

**Pull request closed without merging**